packages/build/src/plugins_core/deploy_config/index.ts

```diff
@@ -0,0 +1,87 @@
+import { promises as fs } from 'fs'
+import { resolve } from 'path'
+
+import merge from 'deepmerge'
+
+import type { NetlifyConfig } from '../../index.js'
+import { CoreStep, CoreStepFunction } from '../types.js'
+
+import { filterConfig } from './util.js'
+
+const ALLOWED_PROPERTIES = [
+  ['build', 'edge_functions'],
+  ['build', 'functions'],
+  ['build', 'publish'],
+  ['functions', '*'],
+  ['functions', '*', '*'],
+  ['headers'],
+  ['images', 'remote_images'],
+  ['redirects'],
+]
+
+const OVERRIDE_PROPERTIES = new Set(['headers!', 'redirects!'])
+
+const coreStep: CoreStepFunction = async function ({
+  buildDir,
+  netlifyConfig,
+  systemLog = () => {
+    // no-op
+  },
+}) {
+  const configPath = resolve(buildDir, '.netlify', 'deploy', 'v1', 'config.json')
+
+  let config: Partial<NetlifyConfig> = {}
+
+  try {
+    const data = await fs.readFile(configPath, 'utf8')
+
+    config = JSON.parse(data) as Partial<NetlifyConfig>
+  } catch (err) {
+    // If the file doesn't exist, this is a non-error.
+    if (err.code === 'ENOENT') {
+      return {}
+    }
+
+    systemLog(`Failed to read Deploy Configuration API: ${err.message}`)
+
+    throw new Error('An error occured while processing the platform configurarion defined by your framework.')
+  }
+
+  const configOverrides: Partial<NetlifyConfig> = {}
+
+  for (const key in config) {
+    // If the key uses the special notation for defining mutations that should
+    // take precedence over user-defined properties, extract the canonical
+    // property, set it on a different object, and delete it from the main one.
+    if (OVERRIDE_PROPERTIES.has(key)) {
+      const canonicalKey = key.slice(0, -1)
+
+      configOverrides[canonicalKey] = config[key]
+
+      delete config[key]
+    }
+  }
+
+  // Filtering out any properties that can't be mutated using this API.
+  config = filterConfig(config, [], ALLOWED_PROPERTIES, systemLog)
+
+  // Merging the different configuration sources. The order here is important.
+  // Leftmost elements of the array take precedence.
+  const newConfig = merge.all([configOverrides, netlifyConfig, config])
```

Comment:
can we use the `mergeConfigs` function from `@netlify/config` for that instead of reimplementing this piece here?

Reply:
Done in https://github.com/netlify/build/pull/5509/commits/a50d8d83a7c33f4b06b49e618d723f8f80b88a3d.